docs/design/kuadrant-api-v2.md

```diff
@@ -0,0 +1,178 @@
+* Path/Operation based protection (i.e. authentication and rate limit)
+
+## Proposed Solution
+
+The kuadrant API will be oriented to meet API management requirements. In that context, the kuadrant
+API will be designed around the OpenAPI standard.
+Specifically, [OpenAPI 3.X specification](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md).
+Kuadrant will define the *APIProduct* object class (a CRD in terms of kubernetes) to wrap a set of
+API management configurations on top of a list of customer owened API definitions.
+
+### API Product
+
+**Custom Resource Definition GVK**
+
+Currently, as of kuadrant v0.2.0, APIProduct [Group Version Kind](https://book.kubebuilder.io/cronjob-tutorial/gvks.html)
+is: [apiproducts.v1beta1.networking.kuadrant.io](https://github.com/Kuadrant/kuadrant-controller/blob/v0.2.0/doc/architecture.md#api-product-crd)
+
+Taking into account kuadrant is not providing networking level API, but **API Management** level API,
+the proposed GVK is: **apiproducts.v1alpha1.apim.kuadrant.io**.
+
+**Domains**
+
+APIProduct resource will be configured with a list of destination domains or hostnames to which
+traffic is being sent.
+
+It is the responsability of cluster operators and API owners to forward *îngress* traffic for those
+specific domains to the kuadrant gateway. A kuadrant gateway would be a gateway configured by the
+kuadrant control plane, having as input the API Product resource.
+
+For example, cluster operators can deploy required (cloud provider level) networking resources to
+forward traffic to the cluster ingress gateways. Then, the API owners can create (cluster level)
+networking resources, like kubernetes Ingress or Gateway API HttpRoutes, to route that traffic to
+the kuadrant gateways. In this example only *North/South* traffic was referenced. By design,ç
+it could also be *East/west* traffic. By design, there is nothing preventing from having *East/West*
+traffic being protected by kuadrant. This is mainly because kuadrant is not a networking API.
+
+**Protected API specification**
+
+Kuadrant will support the open standard
+[OpenAPI 3.X specification](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md)
+for the API spec.
+
+Each *API Product* can hold multiple OAS docs references. The *API Product* `spec` field includes references
+in the form of URLs and config map references. The actual content may be included in the `status` field.
+
+Kuadrant will extend the OAS doc leveraging
+[OAS custom extensions](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md#specificationExtensions)
+`x-kuadrant-*`, initially for rate limit purposes. In the future, OAS extensions could be leveraged
+for other API management features, for instance usage plans. More about the kuadrant custom
+extensions in the spec reference section.
+
+In case of lacking one OpenAPI doc, kuadrant will provide tooling for easy generation of OpenAPI docs.
```

Comment:
Should we be doing this? I think we have to end users. 1) I know all about OAS and use them 2) I don't want or care about OpenAPI specs but know all about k8s and want to use the resources

Reply:
I feel we should support both

Reply:
For the second use case, currently kuadrant supported the "PathMatch" option. If the user does not have OAS, it will need to provide some info:
* path match (`GET /toystore/*`) or set of operations (` [ "GET /toystore/dolls", "GET /toystore/egames" ] `)
* K8s service or upstream URL
* Desired authentication
* Desired rate limit 

With that information, my proposal was to have some tooling available to generate OAS. Can be the CLI, online web app, custom resource for that... With the OAS in place, the user can protect with kuadrant.